driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -289,8 +350,9 @@ public static String quote(String id) {
      * @return {@code true} if the given identifier is a known reserved
      * CQL keyword, {@code false} otherwise.
      */
+
     public static boolean isReservedCqlKeyword(String id) {
-        return id != null && RESERVED_KEYWORDS.contains(id);
+        return id != null && IGNORE_CASE_RESERVED_KEYWORDS.contains(ignoreCase(id));
```

Comment:
I agree that a wrapper is cleaner, but since we're trying to squeeze the last bit of performance, I wonder if we should avoid this extra object instantiation for each method call. That's why I suggested a separate method for the hash and a map for the reserved keywords. WDYT?

Reply:
yes, I thought about the extra object allocation. The problem is what in this case:
```java
int hash = caseInsensitiveHash(id);
String keyword = RESERVED_KEYWORDS.get(hash);
```
we also have a allocation - auto-boxing of primitive int. There is a pool of pre-allocated objects in Integer.valueOf ([-128 ; 127]) but in cause of hashes when values are spread across int range there is no large benefit from it. We can use a special map with primitive int keys to avoid it - will be ok to use https://netty.io/4.0/api/io/netty/util/collection/IntObjectMap.html in the driver code (there are other libraries like Trove but Netty is already present as a dependency)?

Regarding:
```java 
String keyword = RESERVED_KEYWORDS.get(hash)
```
Do you suggest here to use 'perfect hash functions' without collisions on the constant set of keywords?

Reply:
> we also have a allocation - auto-boxing of primitive int

Good point.

> Do you suggest here to use 'perfect hash functions' without collisions on the constant set of keywords?

I hadn't completely thought this through, we'd need a multimap (map of lists) to handle collisions. At this point I'm not sure if this is worth exploring.

Reply:
Have implemented a non-wrap option using io.netty.util.collection.IntObjectHashMap.
Also because there is a granular control on comparison logic for this way - used char[] to avoid boundary checks in equals logic (saves about 10 ns).

Results:

```
# JMH version: 1.19
# VM version: JDK 1.8.0_151, VM 25.151-b12
# VM invoker: C:\Program Files\Java\jre1.8.0_151\bin\java.exe
# VM options: <none>
# Warmup: 5 iterations, 1 s each
# Measurement: 30 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 2 threads, will synchronize iterations
# Benchmark mode: Average time, time/op

Benchmark                                                    Mode  Cnt    Score   Error  Units
MetadataBenchmark.equalsWrapper_NonReservedKeywordMixedCase  avgt   30   49.738 ± 1.313  ns/op
MetadataBenchmark.equalsWrapper_NonReservedKeywordUpperCase  avgt   30   59.430 ± 3.230  ns/op
MetadataBenchmark.equalsWrapper_ReservedKeywordMixedCase     avgt   30   75.837 ± 4.706  ns/op
MetadataBenchmark.equalsWrapper_reservedKeywordUpperCase     avgt   30  100.812 ± 2.402  ns/op

MetadataBenchmark.hashCheck_NonReservedKeywordMixedCase      avgt   30   45.817 ± 1.289  ns/op
MetadataBenchmark.hashCheck_NonReservedKeywordUpperCase      avgt   30   47.084 ± 1.760  ns/op
MetadataBenchmark.hashCheck_ReservedKeywordMixedCase         avgt   30   72.546 ± 1.198  ns/op
MetadataBenchmark.hashCheck_reservedKeywordUpperCase         avgt   30   93.308 ± 1.769  ns/op
```

Code:

```java
import io.netty.util.collection.IntObjectHashMap;
import io.netty.util.collection.IntObjectMap;

private static final IntObjectMap<List<char[]>> KEYWORD_HASHES = new IntObjectHashMap<List<char[]>>();
static {
    for (String keyword : RESERVED_KEYWORDS) {
        char[] keywordAsCharArray = keyword.toLowerCase().toCharArray();
        int hash = ignoreCaseHashCode(keyword);
        List<char[]> keywordsPerHash = KEYWORD_HASHES.get(hash);
        if (keywordsPerHash == null) {
            keywordsPerHash = new ArrayList<char[]>();
            keywordsPerHash.add(keywordAsCharArray);
            KEYWORD_HASHES.put(hash, keywordsPerHash);
        } else {
            keywordsPerHash.add(keywordAsCharArray);
        }
    }
}
    
// keyword is expected as a second argument always in low case
private static boolean equalsIgnoreCaseAscii(String str1, char[] str2LowCase) {
    if (str1.length() != str2LowCase.length) return false;

    for (int i = 0; i < str1.length(); i++) {
        char c1 = str1.charAt(i);
        char c2Low = str2LowCase[i];
        if (c1 == c2Low) {
            continue;
        }
        char low1 = toLowerCaseAscii(c1);
        if (low1 == c2Low) {
            continue;
        }
        return false;
    }
    return true;
}

private static char toLowerCaseAscii(char c) {
    if (c >= 65 && c <= 90) { // A-Z
        c = (char) (c ^ 0x20); // convert to lower case
    }
    return c;
}

private static int ignoreCaseHashCode(String str) {
    int hashCode = 17;
    for (int i = 0; i < str.length(); i++) {
        char c = toLowerCaseAscii(str.charAt(i));
        hashCode = 31 * hashCode + c;
    }
    return hashCode;
}

public static boolean isReservedCqlKeyword(String id) {
    if(id == null) {
        return false;
    }
    int hash = ignoreCaseHashCode(id);
    List<char[]> keywords = KEYWORD_HASHES.get(hash);
    if (keywords == null) {
        return false;
    } else {
        for (char[] keyword : keywords) {
            if (equalsIgnoreCaseAscii(id, keyword)) {
                return true;
            }
        }
        return false;
    }
}
```

Reply:
That is the preference for the last 2 options (ignore case wrapper vs hashes in primitive int map from netty-common)?

Reply:
The primitive int map seems a bit faster, let's go with that.
> used char[] to avoid boundary checks in equals logic

Do you mean the checks made by `charAt(i)`?

Reply:
yes, that logic
ok, thank you, I will update the patch with the primitive int map option.